Refactor front-end. Add support for discord.ext, command_config.json and feedback label.

diivm opened the pull request:
solves #nothing :laughing:

diivm commented on main.py:
```diff
@@ -282,7 +309,10 @@ async def on_message(message):
 Retrieves N least-recently-updated PR awaiting your review
 
 `!q <N>`
```
If this PR is merged, we should change this and it's usage to `rq`

diivm commented:
@kunaltyagi

kunaltyagi commented:
The code is quite similar with `review_q` having difference in only `title_suffix` and label. Could be refactored a bit more to reduce duplication

kunaltyagi commented on main.py:
```diff
@@ -186,62 +186,106 @@ async def set_playing(status):
-    5. return them
-    """
-    await set_playing('Finding Issues')
+def get_params(command):
```
While I appreciate the effort, you've split 3 functions based on the command. That's not good.

Merely modifying `request_q` and the way title was worded would have been sufficient. You could've called `request_q` with 2 different labels and it would have worked just fine

diivm replied:
I'll undo the changes if you want, but I wanted to know why exactly it isn't good? Is it a design issue or it'll be difficult to handle later, what problems may arise?

> Merely modifying request_q and the way title was worded would have been sufficient.

That's true, but I noticed code repetition in the random generator too. Plus, I was also thinking if we wanted to add `needs: feedback issues` (and more filters, if needed), it would be better to have a general function of some sort.

So, do I go back to the previous implementation? Maybe you can suggest an even better alternative structure.

kunaltyagi replied:
> Is it a design issue

Yes. The details for each command is split across multiple functions. That's a bad sign in itself.

I didn't notice the repetition in random-generator before, but maybe restructure in the following way:
1. each command has it's own function which supplies the details (issue/pr, author, include/exclude label)
2. call the search result generator
3. transform the search results (aka replace them with details via another API call)
4. perform some operation on the search result (aka: random select, top select, etc.)

2-4 is a pipeline which can be written as a separate function, with the transformer and operations being independent functions. The higher order function are sent to the pipeline in step 1.

Don't follow this blindly, since I didn't spend a lot of time on the code. The code is very young so feel free to experiment with decorators or control inversion or generators or co-routines, etc.

kunaltyagi replied:
Let's get this one going soon if possible

diivm replied:
Sure.

diivm replied:
> each command has it's own function which supplies the details (issue/pr, author, include/exclude label)

For this, I was thinking of treating it as configurations (dictionaries). Can I do the configurations in a separate file? If yes, I assume a `.json`

kunaltyagi replied:
Converting "Sorting by author" an option in the same command (or derivating using prefix/suffix, additional term, etc.) would be nice too

diivm replied:
Do you mean for the `needs: feedback` label, sort by the PR authors' names as an additional option, on top of the `order by oldest`?

kunaltyagi replied:
I meant something similar to the difference between the query and review commands

diivm replied:
But a PR doesn't need the feedback of an individual, right? If yes, then I guess those will be the reviewers.

kunaltyagi replied:
Oh yeah. Scrap that thought. My bad

diivm commented:
https://github.com/PointCloudLibrary/discord-bot/pull/5/commits/825645a88add7e482655093fd52181a54441c519: How does this class look? 
https://github.com/PointCloudLibrary/discord-bot/blob/825645a88add7e482655093fd52181a54441c519/main.py#L189-L273


If it is good, I'll take that and refactor some more later in the file:
https://github.com/PointCloudLibrary/discord-bot/blob/825645a88add7e482655093fd52181a54441c519/main.py#L316-L379

kunaltyagi commented on command_config.json:
```diff
@@ -0,0 +1,52 @@
+{
+	"type":{		
```
Indent properly please. Use 2 spaces if indentation gets too deep.

SergioRAgostinho replied:
Just format everything through black (python package).

diivm replied:
Sure, that's the plan #7

diivm replied:
Done.

diivm replied:
Note: black adds extra trailing comma while formatting JSON, which leads to errors while parsing using json in python.

kunaltyagi commented on main.py:
```diff
@@ -186,67 +186,91 @@ async def set_playing(status):
+
+    
+    def issues_generator(self):
+        self.issues=[x async for x in get_issues(self.git_config["repository"], self.git_config["closed"],
```
* Have sensible defaults to reduce data in config file
* Use `**` to expand a dictionary for kwargs
```python
def f(hi="what"):
    """
    >>> f(**{"hi":"wert"})
    wert
    """
    print(hi)
```

diivm replied:
Adds a lot of arguments:
```
def issues_generator(repository, closed, pull_request, include_labels, 
                     exclude_labels, sort, ascending_order):
    self.issues=[x async for x in get_issues(repository, closed, pull_request, 
                                             include_labels, exclude_labels,
                                             sort, ascending_order)]
```
```
issues_generator(**self.git_config)
```

kunaltyagi replied:
No. It'll be something similar to
```python
def issue_generator(self):
    self.issues=[x async for x in get_issues(**self.config)]
```

diivm replied:
My bad. 
Although we should default to Python>3.6 since it guarantees order preservation in kwargs. https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-pep468

kunaltyagi replied:
Not needed if repository is also made a keyword argument with a default value of "github/training-kit" since then all arguments will be keyword arguments

diivm replied:
Okay. I'll add the default value in `get_issues`

kunaltyagi commented on command_config.json:
```diff
@@ -0,0 +1,52 @@
+{
+    "type": {
```
`type` isn't needed since all are commands

kunaltyagi commented on command_config.json:
```diff
@@ -0,0 +1,52 @@
+        "rand_issue": {
+            "git": {
+                "repository": "config['repo']",
+                "closed": "False",
```
closed and repository can be safely eliminated from config (and defaulted to the common value). You can keep them in the python though

kunaltyagi commented on command_config.json:
```diff
@@ -0,0 +1,52 @@
+            },
+            "discord": {
+                "game_name": "'On The Cue'",
+                "title": "f'Oldest {number_of_issues} PR' + (f' for @{author}:' if author else ' in review queue:')",
```
That's bad. Don't use f-strings here. Use normal strings, and the `.format()` on a string. Supplying extra variables in format doesn't affect the output.

```python
title = config['discord']['title'].format(number_of_issues=number_of_issues,
    author=author)
```

diivm replied:
Much needed. :+1:

kunaltyagi commented on main.py:
```diff
@@ -186,67 +186,91 @@ async def set_playing(status):
-                            error_channel=error_channel)
-        reply = discord.Embed(color=discord.Color.purple())
-
+class SendReply:
```
Try to have something like:
```py
# derive from a nice class
class Bot(discord.bot):
    def __init__(self):
        config = read_config(file_name)
        commands = [name: make_command(args) for name, args in config.iteritems()]
    def run(name, value):
        command = commands.get(user_input, lambda x: None) # or throw error, etc.
        command(value)
```

diivm replied:
Give me some time, I'll rethink the whole structure

diivm replied:
Would it be better to use [discord.ext](https://discordpy.readthedocs.io/en/latest/ext/commands/)?

kunaltyagi replied:
Yeah, my example and the examples there look similar

diivm replied:
I was planning on treating commands as [extensions](https://discordpy.readthedocs.io/en/latest/ext/commands/extensions.html#extensions). 
A separate extension for each command or all-in-one?

PS: By commands, I mean feedback queue, review queue etc.


EDIT: I am thinking only one extension would be good, with commands arranged 
 separately. ([cogs](https://discordpy.readthedocs.io/en/latest/ext/commands/cogs.html#cogs))

kunaltyagi replied:
Just commands.command decoration might do the trick, without needing the breadth of extensions and cogs.

diivm replied:
Yes, but it becomes neat and arranged that way. 
Plus, the LOC reduces in the main file (I was planning to move that code to a new file since it is an extension).

diivm replied:
I have a doubt. What exactly is `error_handler` doing?
https://github.com/PointCloudLibrary/discord-bot/blob/c27394669f1564df397091286532713262f64be0/main.py#L73-L84

While using [exceptions](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#exceptions), I think it will be safe to remove it (or convert it):
```
@bot.on_command_error
async def error_handler(ctx, error):
    reply = discord.Embed(color=discord.Color.purple())
    reply.description = "Talking to me? Use `!what` to know more."
    if isinstance(error, discord.DiscordException):
        await ctx.send(embed=reply)
```

kunaltyagi replied:
> What exactly is `error_handler` doing?

Timeout, API limit hit issues

kunaltyagi replied:
> the LOC reduces in the main file
> I think it will be safe to remove it (or convert it):

Let's see the prototype and then discuss over it.

diivm replied:
There might be a problem. We shouldn't share variables in async methods, right? Thus, the whole class thing shouldn't be done. If we really needed that, then we need to make methods synchronous, which involves semaphores, etc. (state variables). (async is threaded afaik)

If that's the case, your original code be should be left as it is, i.e. functional in nature and just use discord.ext to clean some message stuff, make commands, etc.

> Let's see the prototype and then discuss over it.

Sorry for the delay, but the above doubt needs clearing.

***

Edit: This would have been perfect but we cannot decorate an instance of a class: 

```diff
- @Command(cmd_name). decorated
- async def sorter1(self):
-     use_self_variables
-     ...
```
Only this:
```
@Command.decorated
async def sorter1():
    no_self_variables
    ...
```

kunaltyagi replied:
https://github.com/Rapptz/discord.py/blob/master/examples/basic_voice.py

kunaltyagi commented:
What we currently need/use the bot for:
* Random issues for cleanup
* feedback label:
  * issue
  * PR
* code review label:
  * general
  * **my** code review

* Reorganize the front end to have three nouns (rand, feedback, review)
* Add a classifier to the noun (issue/pr) so it becomes: `rand 5 pr`, `review 4`, `feedback 3 issue`
  * default to issue for random and feedback
  * not for review, since review is for PR only
  * autocomplete using substring match (issue, issues, is, pr, pull, pulls, pull-requests)
* Either add a classifier to review (general/mine) or just get both the data (general, and sorted by author's name, validated by API. Eg: [for me](https://api.github.com/users/kunaltyagi))

This keeps the commands to a manageable number, reduces the paths needed for current requirement.

diivm committed:
Refactor. Add feedback label, commands

- add command support from discord.ext
- add feedback support
- add noun support

diivm committed:
add config for git params

diivm commented:
On top of my head:

- Not done classes for commands as they are good only if we use cogs
- decorators were creating complexity in understanding the flow of code, would much rather use fn calls to choose issues
  - `oneshot` requires random sampler, so handling everything in a @bot.command would not have worked since params are needed.
- removed discord config, was unnecessary
- added noun support, it modifies config argument according to noun
  - issue: pull_request=False
  - PR: pull_request=True, for the given config

Basically I eliminated `message` event and its argument parsing, and used commands instead, meanwhile converting repeating code into checks and functions, on top of the config changes.

That's all I can remember, it's 03:15 :sleepy:

diivm commented on main.py:
```diff
@@ -183,156 +184,154 @@ def compose_message(issues):
-                    break
-
-        reply.title = f"Oldest {number_of_issues} PR" + title_suffix
+        if git_config["command"]=="rand":
+            chosen_issues=choose_issues_rand(issues, number_of_issues)
+            reply.title = f"{number_of_issues} random picks out of {len(issues)}:"
+        elif git_config["command"]=="review" or git_config["command"]=="feedback":
+            chosen_issues=choose_issues_queue(issues, number_of_issues, author)
+            reply.title = f"Oldest {number_of_issues} " + ("PR" if git_config["pull_request"] else "issues") + (f' for @{author}:' if author else ' in queue:')
```
Do suggest a better alternative, if possible

kunaltyagi replied:
```suggestion
           kind = "PR" if git_config["pull_request"] else "issues"
           selection = f for @{author}" if author else "in queue"
           reply.title = f"Oldest {number_of_issues} {kind} {selection}:"
```
When in doubt, divide (this is only for one line). For the rest, do similar. Divide and conquer

diivm replied:
What about the function `choose_issues_rand`, `choose_issues_queue` and their calls?

kunaltyagi replied:
I'll take another look sometime soon

diivm commented on main.py:
```diff
@@ -183,156 +184,154 @@ def compose_message(issues):
+    await queue_handler(channel=bot.channel, number_of_issues=number_of_issues, author=None, git_config=git_config)
+
+
+@bot.command(name='feedback')
+@bot.command(name='fq')
```
We need to test this

kunaltyagi replied:
Feel free to test it. You need to create a [new application](https://discord.com/developers/)

diivm replied:
On it.

kunaltyagi commented on main.py:
```diff
@@ -183,156 +184,154 @@ def compose_message(issues):
+@bot.command(name='rand')
+async def rand_cmd(ctx, number_of_issues, noun, *):
+    await check_length_issues(number_of_issues)
+    git_config=read_config("command_config.json")["rand"]
```
Let's read this config just once.

diivm replied:
Read once done.
But we need different command_configs 'cause they are updated acc to the noun.

kunaltyagi replied:
Yeah, no issues with modifying copies of command_config

kunaltyagi commented on main.py:
```diff
@@ -183,156 +184,154 @@ def compose_message(issues):
-    args = query[1:]
+async def check_length_issues(number_of_issues):
+    if number_of_issues < 1:
+        reply.description = "I can't give you un-natural issues." + \
```
What's reply?

diivm replied:
added `reply = discord.Embed(color=discord.Color.purple())`

kunaltyagi reviewed (commented):
partial review. Already looks a lot better 😄

kunaltyagi commented on main.py:
```diff
@@ -42,14 +42,16 @@
 
-client = discord.Client()
+bot = commands.Bot(command_prefix='!')
+git_config=read_config("command_config.json")
```
No... it'd be `command_config={}` and read from file happens in one place

diivm replied:
read happening in `modify_params` now

diivm replied:
Done.

read happening in `modify_params` and `one_shot` (cause `one_shot` doesn't call `modify_params`)

kunaltyagi replied:
Why not read in the main, along with git and discord secrets?

diivm replied:
we need command based reading:
 `command_config=read_config("command_config.json")[command_type]
`

kunaltyagi replied:
Why do you need to read the file again and again? 😖

diivm replied:
Apologies.

kunaltyagi commented on main.py:
```diff
@@ -183,156 +185,158 @@ def compose_message(issues):
+        async for issue in issues:
+            chosen_issues.append(issue)
+            if len(chosen_issues) == number_of_issues:
+                break
```
```suggestion
    # since async islice doesn't exist
    async for issue in issues:
        chosen_issues.append(issue)
        if len(chosen_issues) == number_of_issues:
            break
```

diivm replied:
done

kunaltyagi commented on main.py:
```diff
@@ -183,156 +185,158 @@ def compose_message(issues):
+        if command_config["command"]=="rand":
+            chosen_issues=choose_issues_rand(issues, number_of_issues)
+            reply.title = f"{number_of_issues} random picks out of {len(issues)}:"
+        elif command_config["command"]=="review" or command_config["command"]=="feedback":
```
```suggestion
        elif command_config["command"] in ["review", "feedback"]:
```

diivm replied:
done

kunaltyagi commented on main.py:
```diff
@@ -183,156 +185,158 @@ def compose_message(issues):
+        number_of_issues = 10
+        reply.description = "Let's curb that enthusiasm.. just a little"
+        await channel.send(embed=reply)
+    return
```
Does the modification of input variable `number_of_issues` also modify it in the calling function?

diivm replied:
Yeah, pass by reference is the default in Python

[SO](https://stackoverflow.com/questions/986006/how-do-i-pass-a-variable-by-reference)

kunaltyagi replied:
That's not what python does. (assigning an integer doesn't reset the original reference). Python variables are references, but passed by value.

[Repl.it](https://repl.it/repls/GentleMuddyDatalog)

```python
def modify(num):
    num = 10

x = 8;
modify(x)
assert(x == 8)
```
equivalent c++ program
```cpp
#include <memory>

void modify(std::shared_ptr<int> num) { num = std::make_shared(10); }

int main() {
    auto x = std::make_shared(8);
    modify(x);
    assert(x == 8);
    return 0;
}
```

diivm replied:
Yep, a blunder.
`int` is immutable

diivm replied:
Everything now in `modify_params`

kunaltyagi commented on main.py:
```diff
@@ -183,156 +185,158 @@ def compose_message(issues):
+async def rand_cmd(ctx, number_of_issues, noun, *):
+    await check_length_issues(number_of_issues)
+    command_config=git_config["rand"]
+    update_config(noun, command_config)
```
These 2 lines can be merged in 1, if the variable was named better 😄

diivm replied:
`modify_params`

kunaltyagi commented on main.py:
```diff
@@ -183,156 +185,158 @@ def compose_message(issues):
+@bot.command(name='review')
+async def review_cmd(ctx, number_of_issues, noun, *):
+    await check_length_issues(number_of_issues)
+    command_config=git_config["review"]
```
Is there a way to take the name of the command from the decorator itself?

diivm replied:
https://discordpy.readthedocs.io/en/latest/ext/commands/api.html?highlight=command%20name#discord.ext.commands.Command.name

Like this?

kunaltyagi replied:
That wouldn't work. Ignore

kunaltyagi commented on main.py:
```diff
@@ -183,156 +185,158 @@ def compose_message(issues):
+
+
+@bot.command(name='rq'):
+async def rq_cmd(ctx, number_of_issues, noun, *):
```
Perhaps, not have unlimited arguments as (valid) input

diivm replied:
Was an error anyway. Removed.

kunaltyagi commented on main.py:
```diff
@@ -183,156 +185,158 @@ def compose_message(issues):
+    command_config=git_config["feedback"]
+    update_config(noun, command_config)
+
+    await queue_handler(channel=bot.channel, number_of_issues=number_of_issues, author=None, command_config=command_config)
```
Let's remove the command details from `queue_handler`, put them in the commands and just keep the message sending logic there. For one-shot, we can make another function which is called in both contexts

diivm replied:
There isn't much of message sending logic. It defies the initial goal of code duplication reduction, since much of the code is same for all commands.

kunaltyagi commented on command_config.json:
```diff
@@ -0,0 +1,26 @@
+{
+    "rand": {
+        "command": "rand",
```
Goal: We shouldn't need `command` here

diivm replied:
yep, not suitable for `issues = [x async for x in get_issues(**command_config)]` anyway

Done.

diivm commented:
@kunaltyagi 
I would look into the suggestions after the current code (corrected for silly errors) work for the testing application I am running.
Some of the changes you mentioned have already been addressed while debugging.

Current status:
After making some changes, the one-shot mode is running fine but the `choose_issues` function is not working as expected (returning `There weren't enough` always, i.e. choose_issues list is not properly being formed)
Also some weird error about `async for` is invalid syntax. Probably some version issue, it's a keyword maybe.

After it gets working, I'll address the remaining changes.

Edit: `async for` worked, one-shot worked

kunaltyagi commented:
I've made this a draft. Ping me when you think I should take a look 😄

diivm commented on main.py:
```diff
@@ -384,15 +392,15 @@ def main():
-    loop.create_task(oneshot(args.channel_id, args.issues))
-    loop.run_until_complete(client.connect())
+    loop.run_until_complete(bot.login(config["discord_token"]))
+    # loop.create_task(oneshot(args.channel_id, args.issues))
```
testing purposes

diivm commented:
@kunaltyagi 
It's running fine as of this moment :laughing: 
Ready for further discussion, I guess.

- Clean rebase after final approval
- Complete formatting using black in new PR.

kunaltyagi commented on main.py:
```diff
@@ -183,167 +190,168 @@ def compose_message(issues):
+    await queue_handler(command_type="feedback", channel=ctx.channel, number_of_issues=number_of_issues, author=None, command_config=command_config)
+
+
+@bot.command(name='fq')
```
The inverted splitting of logic (flow goes into one function only to be split again) is present still, though the code is cleaner than before.

I'm writing a prototype of what can be done, not necessarily should be done:
```python
def setup(bot, config):
    for name, functor in config["commands"].items():
        def magic_function(config["cmd_config"][name]):
            async with (): # from queu_handler
                setup(magic_variables)  # aka everything before await queue_handler
                get_issues()  # line 200 or so in queue_handler
                functor(magic_variables)
                send_message() # aka the ending of queue_handler

        bot.command(name=name)(magic_function)
```

This function can be itself made a decorator like `@bot.command` since decorators are just syntactic sugar

diivm replied:
Some pointers:
- ~~The parameters have to be specified for each function that calls the decorator.~~
  ```
  @bot.command(name='rand')
  async def rand_cmd(ctx, number_of_issues: int, noun):
  ```
  ```
  @bot.command(name='review')
  async def review_cmd(ctx, number_of_issues: int):
  ```
  ```
  @bot.command(name='what')
  async def what_cmd(ctx):
  ```
  ~~The functions require different arguments which have to be specified manually or through if-else while defining the `bot.command(name=name)`~~ Can be handled

- ~~`bot.command(name=name)` means we have to duplicate `review` and `rq`, `feedback` and `fq` in the config file or if-else handling.~~

- I tried putting 
```
            if author:
                issues = pr_with_pending_review(get_pr_details(issues), author)
                # since async islice doesn't exist
                async for issue in issues:
                    chosen_issues.append(issue)
                    if len(chosen_issues) == number_of_issues:
                        break
            else:
                chosen_issues=issues[:number_of_issues]
```
in separate async function, it didn't work so well (it wasn't updating the `chosen_issues` list, giving out no results. It should update in a `def` fn, but there is something about async that was causing this issue. Still don't know the cause.
Help will be appreciated.

- ~~The `if command_type=="rand":` and `if command_type in ["review", "feedback"]:` still need to be handled for this.~~

diivm replied:
Opinion needed:
- Remove `fq` and `rq` options, and use `me` as the noun for author specific review PRs. 

`fq` and `feedback` are same, and `rq` is just a differentiator for author, can instead use a noun, as review can only be done for PRs, thus also eliminating the noun, no-noun problem.

Will also solve the duplicacy problem for command_config.json

kunaltyagi replied:
Invert the logic. `all` for not filtering, else filtered by author.

diivm replied:
It ain't working :weary:  
It theoretically should. Will look more into it.
Wish there were preprocessor macros in python.

diivm replied:
The issue is with the loop variables (name, conf), probably [this](https://stackoverflow.com/a/3431699)

I can access `name` anywhere, but not `conf` since `conf=conf` isn't done. Printing `conf` inside the function declaration gives no result, while outside, it can print.

kunaltyagi replied:
Put things in a class, and don't run the loop in global level, but in a function. That'd resolve the issue (i think)

diivm replied:
That's solved, forgot to write. Problem was late binding as mentioned [here](https://stackoverflow.com/questions/3431676/creating-functions-in-a-loop/3431699#3431699). 
Solved by just passing `name=name` and `conf=conf` extra params to `magic_function`.

Great solution btw (defining functions in a loop) :heart:

diivm replied:
The `async for issue in issues` is still troubling, maybe it doesn't work well in `if` clauses

kunaltyagi replied:
Try using `for issue in issues` 🤷‍♂️

kunaltyagi replied:
PS: I'd recommend using `partial function` instead of eager bindings since that'd be explicitly informing the reader about the issue. See this https://docs.python-guide.org/writing/gotchas/#late-binding-closures

diivm replied:
> Try using for issue in issues

AsyncGenerator, can't

diivm replied:
> partial function

"Return a new partial object which when called will behave like func called with the positional arguments args and keyword arguments keywords."
We don't want an object, rather a function definition.

This doesn't work:
```
    bot.command(name=name)(magic_function(ctx, int(number_of_issues), noun))

```

Plus anyway `name` and `conf` are required in the function.

diivm commented on main.py:
```diff
@@ -183,167 +190,120 @@ def compose_message(issues):
+                if author:
+                    issues = pr_with_pending_review(get_pr_details(issues), author)
+                    # since async islice doesn't exist
+                    async for issue in issues:
+                        chosen_issues.append(issue)
+                        if len(chosen_issues) == number_of_issues:
+                            break
```
still troubling.

kunaltyagi replied:
What's the issue exactly?

diivm replied:
Everything works, except for personal review. For me, it should generate `There weren't enough`. But it gives no reply.
It fetches issues but the control doesn't go in this loop, due to which `chosen_issues` isn't updated.

`async for` isn't being executed, it gives no errors,  I observed similar behaviour while debugging when some other async thing wasn't working.

The code is used as a workaround for `async islice`, `issues` is an AsyncGenerator.

In the original code, it works fine, which is weird because it should work here too.

kunaltyagi replied:
1. `except for personal review`: Personal feedback is working?
2. Is control reaching L261?
3. Is `author == None`?

diivm replied:
1. No such thing as personal feedback, sir. :laughing: 
2. Nope, that's the issue
3. Nope, if author==None, i.e for general review (`all`) and feedback case, it works fine (gives prs/issues for the number specified). L267 is executed.
   Edit: Additionally, it prints the author name in the `if author:` loop, so author isn't None

kunaltyagi replied:
Check if the generator is empty.

diivm replied:
We need to access it via `async for` loop, which isn't working. Is there any other method?

kunaltyagi replied:
Hop over to discord

kunaltyagi commented on main.py:
```diff
@@ -255,7 +256,7 @@ async def magic_function(ctx, number_of_issues: int, noun, name=name, conf=conf)
                 reply.title = f"{number_of_issues} random picks out of {len(issues)}:"
 
-            elif name in ["review", "feedback"]:
+            elif name=="review":
```
What happens for `feedback`?

diivm replied:
Corrected that, committing

kunaltyagi commented on main.py:
```diff
@@ -198,7 +198,9 @@ async def on_ready():
 
-async def modify_params(ctx, conf, name, number_of_issues, noun=None, error_channel=default_error_handler):
+async def modify_params(ctx, conf, name, number_of_issues, noun=None, channel=None, error_channel=default_error_handler):
+    if channel is None:
```
This one shouldn't be needed.

diivm replied:
Sure.

diivm commented on main.py:
```diff
@@ -183,167 +190,121 @@ def compose_message(issues):
+Retrieves N least-recently-updated PR(s) awaiting your review [awaiting general review]
 
-`!q <N>`
-Retrieves N least-recently-updated PR in the review queue'''
-        await channel.send(embed=reply)
-        return
+`!feedback <N> issue/pr`
+Retrieves N least-recently-updated issue(s)/PR(s) in the feedback queue'''
```
@kunaltyagi 
Do you want to change these commands to `rq` and `fq`?

diivm commented:
I guess it's enough for this PR, more for the bot later.
Let me know if anything to be done prior to merge, apart from formatting.

kunaltyagi commented:
Can you update the PR description?

diivm commented:
@kunaltyagi 
Should I format using `black` in this PR itself?

kunaltyagi commented on main.py:
```diff
@@ -183,167 +190,122 @@ def compose_message(issues):
+                author = author.name
+    else:
+        if noun in ["issue", "issues", "ISSUE", "ISSUES"]:
+            conf["pull_request"]=False
```
Seeing this, does `pull_request` need to be there in the config file?

diivm replied:
Good point. I'll remove that and use it as a variable in the code itself.

kunaltyagi commented on main.py:
```diff
@@ -183,167 +190,122 @@ def compose_message(issues):
+        number_of_issues = 10
+        await error_channel("Woah, there!", "Let's curb that enthusiasm.. just a little")
+    
+    if name=="rq":
```
Move this check down (aka to `magic_function`)

diivm replied:
Check nouns in two different functions?

diivm replied:
Maybe I'll split those checks and do something similar like you recommended below https://github.com/PointCloudLibrary/discord-bot/pull/5#discussion_r426139665

diivm replied:
How does bd2e093 look?

I wanted to avoid the `if-else` in [**L287-294**](https://github.com/PointCloudLibrary/discord-bot/commit/bd2e093c58f65a87fc12c87a34abce4af884bd3d#diff-5bc02cefb3ea9e27f1a6776eabd1935dR287-R294)

Better, but can't await `None`, `True`:
```
        check_for_commands={"rand": [check_order[0], check_order[1], None], "rq": [check_order[0], True, check_order[1]], "fq": [check_order[0], check_order[1], None]}
        
        number_of_issues=await check_for_commands[name][0]
        pull_request=await check_for_commands[name][1]
        author=await check_for_commands[name][2]
```
Possible fix for the above, but it awaits in the dict itself, so no good:
```
        check_order=[check_number_of_issues(number_of_issues=number_of_issues, error_channel=error_channel), check_pull_request(noun=noun, error_channel=error_channel),  check_author(ctx=ctx, noun=noun, error_channel=error_channel)]
        check_for_commands={"rand": [await check_order[0], await check_order[1], None], "rq": [await check_order[0], True, await check_order[1]], "fq": [await check_order[0], await check_order[1], None]}
        
        number_of_issues=check_for_commands[name][0]
        pull_request=check_for_commands[name][1]
        author=check_for_commands[name][2]
```

Basically, something like [L300-301](https://github.com/PointCloudLibrary/discord-bot/pull/5/commits/bd2e093c58f65a87fc12c87a34abce4af884bd3d#diff-5bc02cefb3ea9e27f1a6776eabd1935dR300-R301)

kunaltyagi commented on main.py:
```diff
@@ -183,167 +190,122 @@ def compose_message(issues):
+
+for name, conf in command_config.items():
+    @bot.command(name=name)
+    async def magic_function(ctx, number_of_issues: int, noun=None, channel=None, name=name, conf=conf): # name=name and conf=conf used to prevent late binding
```
You do a better name than `magic_function` 😆

diivm replied:
:disappointed:

kunaltyagi commented on main.py:
```diff
@@ -183,167 +190,122 @@ def compose_message(issues):
+            reply = discord.Embed(color=discord.Color.purple())
+
+            chosen_issues=[]
+            if name=="rand":
+                """
+                @TODO: improve algorithm to save queries
+                1. get first batch of github_max (100), find total number
+                2. if total < requested, return all
+                3. Generate requested random numbers
+                4. get issues till max(generated_list)
+                5. return them
+                """
+                chosen_issues = random.choices(issues, k=number_of_issues)
+                reply.title = f"{number_of_issues} random picks out of {len(issues)}:"
+
+            elif name in ["rq", "fq"]:
+                # the author check is only for review, author=None for feedback
+                if author:
+                    issues = pr_with_pending_review(get_pr_details(issues), author)
+                    # since async islice doesn't exist
+                    async for issue in issues:
+                        chosen_issues.append(issue)
+                        if len(chosen_issues) == number_of_issues:
+                            break
+                else:
+                    chosen_issues=issues[:number_of_issues]
+                kind = "PR(s)" if conf["pull_request"] else "issue(s)"
+                queue_type = "review" if name=="rq" else "feedback"
+                selection = f"for @{author}" if author else f"in {queue_type} queue"
+                reply.title = f"Oldest {number_of_issues} {kind} {selection}:"
+            
```
Some minor refactoring so we can do the following:
```suggestion
            chosen_issues, reply = command_function[name](author, issues, reply)
```

Reason: deep indentations, long function length

diivm replied:
See bd2e093
It'd be better for the choosing parts to be in different functions, rather if-else, as pointed out by you earlier.

kunaltyagi commented on main.py:
```diff
@@ -183,167 +190,122 @@ def compose_message(issues):
-`!review <N>`
-Retrieves N least-recently-updated PR awaiting your review
+`!rq <N> [all]`
+Retrieves N least-recently-updated PR(s) awaiting your review [awaiting general review]
```
Slightly confusing description. Can some rewording be done to make it better? Not a blocker.

diivm replied:
Can you make a suggestion?

kunaltyagi replied:
Retrieves least-recently-updated PR(s) and filters those awaiting a review from you (default) or anyone (in presence of `all`)

kunaltyagi reviewed (commented):
I'd suggest black here, because you're touching almost all lines anyways. Another option:
* Separate PR for black, let that merge first
* Same config here

kunaltyagi commented on main.py:
```diff
@@ -195,155 +240,107 @@ async def give_random(channel, number_of_issues):
-    await give_random(client.get_channel(channel_id), n)
-    await client.close()
+    await bot.wait_until_ready()
+    await command_function(ctx=None, number_of_issues=n, noun="issue", channel=bot.get_channel(channel_id), name="rand", conf=read_config("command_config.json")["rand"])
```
The filename should appear once only. Also, wouldn't the global variable be read by the time interpreter comes down till here?

diivm replied:
It will be. Removed.

kunaltyagi commented on main.py:
```diff
@@ -384,15 +381,15 @@ def main():
-    loop.create_task(oneshot(args.channel_id, args.issues))
-    loop.run_until_complete(client.connect())
+    loop.run_until_complete(bot.login(config["discord_token"]))
+    # loop.create_task(oneshot(args.channel_id, args.issues))
```
Remember to undo this

kunaltyagi commented on main.py:
```diff
@@ -195,155 +240,107 @@ async def give_random(channel, number_of_issues):
-def read_config(filename):
-    with open(filename, "r") as f:
-        return json.load(f)
+@bot.command(name="review")
```
Is this the Deprecation policy for bot?

diivm replied:
You tell me :smile:

kunaltyagi reviewed (commented):
Looks very good to me

kunaltyagi commented:
Squash as needed and ping back. Thanks for the effort you've put in this @divmadan

diivm commented:
@kunaltyagi 
1. #8 
2. https://github.com/PointCloudLibrary/discord-bot/pull/5#pullrequestreview-413064651
3.  :rocket: 

Thanks for all the help!

kunaltyagi commented:
Conflicts here

diivm commented:
Good to go!

kunaltyagi reviewed (commented):
I've not testing this locally, but the code (in theory) looks good. Will be back after testing.
